docs/null-mapping.md

```diff
@@ -0,0 +1,125 @@
+
+- String and binary datatypes are copied into a mixed list of char or byte lists 
+
+- Nested datatypes (list, map, struct, union, dictionaries) are represented by a mixed list of sublists, depending on the child datatypes (using recursion to populate the child lists)
+  
+  Full details are provided here https://code.kx.com/q/interfaces/arrow/arrow-types/
+
+An arrow table is a set of such arrays, which arrowkdb converts to a mixed list of lists, one per column (although in the case of the nested datatypes one column can be represented by a further set of lists).  By decorating with the field names from the schema, this mixed list of array data then becomes a kdb table.
+
+# 
+
+# Mapping arrow nulls to kdb nulls
+
+## Approach
+
+Currently the simple datatype arrays are memcpy-ed like:
+
+```cpp
+  case arrow::Type::UINT16:
+  {
+    auto uint16_array = std::static_pointer_cast<arrow::UInt16Array>(array_data);
+    memcpy(kH(k_array), uint16_array->raw_values(), uint16_array->length() * sizeof(arrow::UInt16Array::value_type));
+    break;
+  }
+```
+
+A simple approach to map arrow nulls to kdb nulls is to change this to:
+
+```cpp
+  case arrow::Type::INT16: 
+  {
+    auto int16_array = std::static_pointer_cast<arrow::Int16Array>(array_data);
+
+    for (auto i = 0; i < int16_array->length(); ++i)
+      if (int16_array->IsNull(i))
+        kH(k_array)[i] = INT16_MIN;
+      else
+        kH(k_array)[i] = int16_array->Value(i);
+    break;
+  }
+```
+
+The issue with this is that it would result in a significant drop in performance due to inevitable failures in branch prediction.  However, with some arithmetic trickery the same functionality can be modelled without a branch:
+
+```cpp
+  case arrow::Type::INT16:
+  {
+    auto int16_array = std::static_pointer_cast<arrow::Int16Array>(array_data);
+    for (auto i = 0; i < int16_array->length(); ++i)
+      kH(k_array)[i] = (int16_array->IsNull(i) * INT16_MIN) + (!int16_array->IsNull(i) * int16_array->Value(i));
+    break;
+  }
+```
+
+Although there would still be a loss of performance (memcpy copies 64 bits at a time so doing an item by item copy where the datatype < 64 bits will be slower, plus the overhead of indexing into the null bitmap), that loss should not be significant. 
+
+Note: the examples above refer to the reader functions but similar functionality would be provided for the writer functions which would perform the reverse operation (setting the arrow null bitmap if the value is a kdb null).
+
+## Considerations
+
+The bigger problem is that unlike arrow, not all kdb types have a null value.  Also, those that do overload one value in the range (typically INT_MIN or FLOAT_MIN).  For example:
+
+- Each item in an arrow boolean array can be 0b, 1b or NULL.  Kdb has no boolean null.  Similarly it doesn't have a byte null. 
+
+- Unlike arrow, kdb can't distinguish between a null string and empty string.  Similarly it can't distinguish between the " " character and null. 
```

Comment:
```suggestion
- Unlike arrow, kdb can't distinguish between a null string and an empty string.  Similarly it can't distinguish between the " " character and null. 
```